File: behavioral/dart/state_structured/audio_player.dart
import 'interfaces.dart';

class AudioPlayer {
  late AudioPlayerState _currentState;
  bool _isPlaying = false;
  int _currentTrack = 1;
  int _totalTracks;
  
  AudioPlayer(this._totalTracks) {
    _currentState = StoppedState(this);
  }
  
  void setState(AudioPlayerState state) {
    _currentState = state;
  }
  
  void clickPlay() {
    _currentState.clickPlay();
  }
  
  void clickStop() {
    _currentState.clickStop();
  }
  
  void clickNext() {
    _currentState.clickNext();
  }
  
  void clickPrevious() {
    _currentState.clickPrevious();
  }
  
  void startPlaying() {
    _isPlaying = true;
    print('Started playing track $_currentTrack');
  }
  
  void stopPlaying() {
    _isPlaying = false;
    print('Stopped playing');
  }
  
  void nextTrack() {
    _currentTrack = (_currentTrack % _totalTracks) + 1;
    print('Moved to track $_currentTrack');
  }
  
  void previousTrack() {
    _currentTrack = _currentTrack == 1 ? _totalTracks : _currentTrack - 1;
    print('Moved to track $_currentTrack');
  }
  
  bool get isPlaying => _isPlaying;
}

class PlayingState implements AudioPlayerState {
  final AudioPlayer _player;
  
  PlayingState(this._player);
  
  @override
  void clickPlay() {
    print('Already playing');
  }
  
  @override
  void clickStop() {
    print('Stopping playback...');
    _player.stopPlaying();
    _player.setState(StoppedState(_player));
  }
  
  @override
  void clickNext() {
    print('Skipping to next track...');
    _player.nextTrack();
  }
  
  @override
  void clickPrevious() {
    print('Going to previous track...');
    _player.previousTrack();
  }
}

class StoppedState implements AudioPlayerState {
  final AudioPlayer _player;
  
  StoppedState(this._player);
  
  @override
  void clickPlay() {
    print('Starting playback...');
    _player.startPlaying();
    _player.setState(PlayingState(_player));
  }
  
  @override
  void clickStop() {
    print('Already stopped');
  }
  
  @override
  void clickNext() {
    print('Skipping to next track...');
    _player.nextTrack();
  }
  
  @override
  void clickPrevious() {
    print('Going to previous track...');
    _player.previousTrack();
  }
}

File: behavioral/typescript/state_structured/audio_player.ts
import { AudioPlayerState } from './interfaces';

export class AudioPlayer {
  private currentState: AudioPlayerState;
  private isPlaying: boolean = false;
  private currentTrack: number = 1;
  private totalTracks: number;
  
  constructor(tracks: number) {
    this.totalTracks = tracks;
    this.currentState = new StoppedState(this);
  }
  
  setState(state: AudioPlayerState): void {
    this.currentState = state;
  }
  
  clickPlay(): void {
    this.currentState.clickPlay();
  }
  
  clickStop(): void {
    this.currentState.clickStop();
  }
  
  clickNext(): void {
    this.currentState.clickNext();
  }
  
  clickPrevious(): void {
    this.currentState.clickPrevious();
  }
  
  startPlaying(): void {
    this.isPlaying = true;
    console.log(`Started playing track ${this.currentTrack}`);
  }
  
  stopPlaying(): void {
    this.isPlaying = false;
    console.log('Stopped playing');
  }
  
  nextTrack(): void {
    this.currentTrack = (this.currentTrack % this.totalTracks) + 1;
    console.log(`Moved to track ${this.currentTrack}`);
  }
  
  previousTrack(): void {
    this.currentTrack = this.currentTrack === 1 ? this.totalTracks : this.currentTrack - 1;
    console.log(`Moved to track ${this.currentTrack}`);
  }
  
  getIsPlaying(): boolean {
    return this.isPlaying;
  }
}

export class PlayingState implements AudioPlayerState {
  private player: AudioPlayer;
  
  constructor(player: AudioPlayer) {
    this.player = player;
  }
  
  clickPlay(): void {
    console.log('Already playing');
  }
  
  clickStop(): void {
    console.log('Stopping playback...');
    this.player.stopPlaying();
    this.player.setState(new StoppedState(this.player));
  }
  
  clickNext(): void {
    console.log('Skipping to next track...');
    this.player.nextTrack();
  }
  
  clickPrevious(): void {
    console.log('Going to previous track...');
    this.player.previousTrack();
  }
}

export class StoppedState implements AudioPlayerState {
  private player: AudioPlayer;
  
  constructor(player: AudioPlayer) {
    this.player = player;
  }
  
  clickPlay(): void {
    console.log('Starting playback...');
    this.player.startPlaying();
    this.player.setState(new PlayingState(this.player));
  }
  
  clickStop(): void {
    console.log('Already stopped');
  }
  
  clickNext(): void {
    console.log('Skipping to next track...');
    this.player.nextTrack();
  }
  
  clickPrevious(): void {
    console.log('Going to previous track...');
    this.player.previousTrack();
  }
}
